pkg/virtctl/version/version_test.go

```diff
@@ -0,0 +1,90 @@
+package version_test
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	goruntime "runtime"
+	"sync"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	virt_version "kubevirt.io/client-go/version"
+	"kubevirt.io/kubevirt/tests"
+
+	"kubevirt.io/client-go/kubecli"
+)
+
+var _ = Describe("Version", func() {
+
+	var ctrl *gomock.Controller
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		kubecli.GetKubevirtClientFromClientConfig = kubecli.GetMockKubevirtClientFromClientConfig
+		kubecli.MockKubevirtClientInstance = kubecli.NewMockKubevirtClient(ctrl)
+		serverVersionInterface := kubecli.NewMockServerVersionInterface(ctrl)
+		kubecli.MockKubevirtClientInstance.EXPECT().ServerVersion().Return(serverVersionInterface).AnyTimes()
+		serverVersionInterface.EXPECT().Get().Return(&virt_version.Info{
+			GitVersion:   "v0.46.1",
+			GitCommit:    "fda30004223b51f9e604276419a2b376652cb5ad",
+			GitTreeState: "clear",
+			BuildDate:    time.Now().Format("%Y-%m-%dT%H:%M:%SZ"),
+			GoVersion:    goruntime.Version(),
+			Compiler:     goruntime.Compiler,
+			Platform:     fmt.Sprintf("%s/%s", goruntime.GOOS, goruntime.GOARCH),
+		}, nil,
+		).AnyTimes()
+	})
+
+	Context("should be print a message if server and client virtctl versions are different", func() {
+		It("in version command", func() {
+			output := captureOutput(func() {
+				cmd := tests.NewVirtctlCommand("version")
+				Expect(cmd.Execute()).To(BeNil())
+			})
+			fmt.Printf(output)
+			Expect(output).To(HavePrefix("You are using a client virtctl version that is different from the server one"))
+		})
+
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+})
+
+func captureOutput(f func()) string {
+	reader, writer, err := os.Pipe()
+	if err != nil {
+		panic(err)
+	}
+	stdout := os.Stdout
+	stderr := os.Stderr
+	defer func() {
+		os.Stdout = stdout
+		os.Stderr = stderr
+		log.SetOutput(os.Stderr)
+	}()
+	os.Stdout = writer
+	os.Stderr = writer
+	log.SetOutput(writer)
+	out := make(chan string)
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go func() {
+		var buf bytes.Buffer
+		wg.Done()
```

Comment:
I don't think you're using the waitgroup correctly, you're calling `wg.Done()` as soon as the goroutine starts which basically means that the goroutine and the main routine critical sections will basically run in parallel.

Anyway I don't think you need goroutines and waitgroups at all, you can just do
```
f()
io.Copy(&buf, reader)
writer.Close()
reader.Close()
return buf.String()
```
Unless I'm missing something this should work (haven't tried it).

Reply:
Thanks. As agreed offline, I'll remove the `captureOutput`, using `cmd.SetOut()` and `cmd.Printf()` instead.